lib/web_ui/lib/src/engine/keyboard.dart

```diff
@@ -132,16 +127,18 @@ class Keyboard {
     };
 
     EnginePlatformDispatcher.instance.invokeOnPlatformMessage('flutter/keyevent',
-        _messageCodec.encodeMessage(eventData), _noopCallback);
-  }
-
-  bool _shouldPreventDefault(html.KeyboardEvent event) {
-    switch (event.key) {
-      case 'Tab':
-        return true;
-      default:
-        return false;
-    }
+      _messageCodec.encodeMessage(eventData), (ByteData? data) {
+        if (data == null) {
+          return;
+        }
+        final String response = utf8.decode(data.buffer.asUint8List());
+        final Map<String, dynamic> jsonResponse = json.decode(response);
+        if (jsonResponse['handled'] as bool) {
+          // If the framework handled it, then don't propagate it any further.
+          event.preventDefault();
+        }
+      },
+    );
```

Comment:
FYI this file is being heavily refactored in https://github.com/flutter/engine/pull/23466.

Reply:
Yeah, I'm quite aware (and Tong knows I'm working on this).